### PR TITLE
Keep original variable names in metadata to improve diagnostics

### PR DESCRIPTION
**Pull request by ehsantn**
Optimizations like copy propagation rename variables, which can make diagnostics and error messages confusing. This PR keeps the original variable names by user in IR metadata to be used when necessary.

### CODE REVIEW

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the patch. I presume this issue is because parfors and copy propagate are not using SSA, in which case the versioned variable name could be queried?

File: numba/parfors/parfor.py [hunk 1]
```diff
@@ -3838,6 +3839,16 @@ def try_fuse(equiv_set, parfor1, parfor2):
     def is_equiv(x, y):
         return x == y or equiv_set.is_equiv(x, y)
 
+    def get_user_varname(v):
+        """get original variable name by user if possible"""
+        if not isinstance(v, ir.Var):
+            return v
+        v = v.name
+        if "var_rename_map" in metadata and v in metadata["var_rename_map"]:
+            user_varname = metadata["var_rename_map"][v]
```
[the remark this answers is not in the document]
Comment (ehsantn) [reply numba/parfors/parfor.py:3851]: Done.

**Comment by ehsantn** [comment]
I don't understand how copy propagation could use SSA. It replaces variables altogether - it's not a versioning process.

**Review by stuartarchibald — DISMISSED** [review]
Thanks for the fix. Looks good.

**Comment by stuartarchibald** [comment]
@ehsantn c6b3404 looks fine, could you merge in mainline please to resolve conflicts then this can be merged. Many thanks.

**Comment by stuartarchibald** [comment]
> I don't understand how copy propagation could use SSA. It replaces variables altogether - it's not a versioning process.

Yep, sorry, my mistake, this is replacing not versioning.

**Comment by ehsantn** [comment]
Thanks. Merged with master.

**Review by stuartarchibald — APPROVED** [review]
Thanks for the patch.

**Pull request merged**